File: packages/Server/Application/Members/MemberService.cs
namespace Schulcast.Application.Members;

public class MemberService : CrudService<Member>
{
	public MemberService(IDatabase database, ICurrentUserService currentUserService) : base(database, currentUserService) { }

	public async Task<Member?> GetAuthenticatedOrDefault()
	{
		return _currentUserService.Id is null ? null : await Get(_currentUserService.Id.Value);
	}

	public async Task<Member> GetAuthenticated()
	{
		return _currentUserService.Id is null
			? throw new UnauthorizedAccessException()
			: await Get(_currentUserService.Id.Value);
	}

	public async Task<Member> GetByNickname(string Nickname)
	{
		return (await _database.Members.FirstAsync(m => m.Nickname == Nickname)).WithoutPassword();
	}

	public override async Task<IEnumerable<Member>> GetAll()
	{
		return (await base.GetAll()).WithoutPasswords().ExcludeSuperAdmin();
	}

	public override async Task<Member> Get(int id)
	{
		return (await base.Get(id)).WithoutPassword();
	}

	public override async Task<Member> Create(Member model)
	{
		return (await base.Create(model)).WithoutPassword();
	}

	public override async Task<Member> Update(Member model)
	{
		if (string.IsNullOrWhiteSpace(model.Password))
		{
			var existingMember = await Get(model.Id);
			model.Password = existingMember.Password;
		}
		return (await base.Update(model)).WithoutPassword();
	}

	public async Task<Member> Authenticate(string nickname, string password)
	{
		if (nickname is null)
		{
			throw new Exception("The username is not provided");
		}

		if (password is null)
		{
			throw new Exception("The password is not provided");
		}

		var member = await GetByNickname(nickname);

		if (member.Password?.ToLower() != Sha256.GetHash(password).ToLower())
		{
			throw new SecurityException("Incorrect Password");
		}

		member.Token = Token.IssueAccountAccess(TimeSpan.FromDays(30), member);
		return member.WithoutPassword();
	}

	public async Task<IEnumerable<Post>> GetMemberPosts(int id)
	{
		var member = await Get(id);
		return member.Posts;
	}
}

File: packages/Client/sdk/services/MemberService.ts
import { Api, Member } from 'sdk'

export class MemberService {
	static get(id: number) {
		return Api.get<Member>(`/member/${id}`)
	}

	static getAuthenticated() {
		return Api.get<Member | undefined>(`/member/authenticated`)
	}

	static getAll() {
		return Api.get<Array<Member>>(`/member`)
	}

	static save(member: Member) {
		return !member.id
			? Api.post<Member>('/member', member)
			: Api.put<Member>(`/member`, member)
	}

	static delete(member: Member) {
		return Api.delete(`/member/${member.id}`)
	}

	static async authenticate(username: string, password: string) {
		const token = await Api.get<string>(`/member/authenticate?nickname=${username}&password=${password}`)
		Api.authenticator?.authenticate(token)
	}

	static async isAuthenticated() {
		// eslint-disable-next-line no-return-await
		return !!Api.authenticator?.isAuthenticated() && await Api.get<boolean>('/member/is-authenticated')
	}

	static unauthenticate() {
		Api.authenticator?.unauthenticate()
	}
}